sycl/doc/extensions/experimental/sycl_ext_oneapi_device_global.asciidoc

```diff
@@ -324,6 +331,37 @@ The object of type `T` is initialized from the `args` parameter pack using list
 
 `T` must be trivially destructible.
 
+// --- ROW BREAK ---
+a|
+[source,c++]
+----
+constexpr device_global(const device_global &other);
+----
+|
+Available if `has_property<device_image_scope> == false`.
+
+Constructs a `device_global` object, and implicit storage for `T` in the global address space on each device that may access it.
+
+The storage on each device for `T` is initialized with a copy of the storage in `other`.
+
+`T` must be copy constructible and trivially destructible.
+
+// --- ROW BREAK ---
+a|
+[source,c++]
+----
+template <typename OtherT, typename OtherProps>
+constexpr device_global(const device_global<OtherT, OtherProps> &other) {}
+----
+|
+Available if `has_property<device_image_scope> == false`.
```

Comment:
Should this now be:
```suggestion
Available if `has_property<device_image_scope> == false` and `other.has_property<device_image_scope> == false`
```

Reply:
It should be fine if `other` has the property. We can copy from its initializer on host without issue.

Reply:
I might not understand `device_image_scope` correctly, but it seems to me like there's an ambiguity there.

If the left-hand-side doesn't have `device_image_scope`, there's one variable with that name for each device.  If the right-hand-side _does_ have `device_image_scope`, each device image has its own variable (and so the name on the right-hand-side effectively references multiple variables from the perspective of the host).  Which device image is used?

Writing this out made me think about another wrinkle here... Does this assignment have to affect all copies of the `device_global` across all devices?

Reply:
The copy ctor can only be used at initialization. No changes can have been made to the `device_global` at that point. The reason why we can't allow LHS being a `device_image_scope` annotated `device_global` is more of an implementation restriction than a logical one. The problem is that our compiler does not currently support more complex ctors on device, which would be required if we allowed the copying, as the kernel would have to run the ctor that copies the initial value.

Reply:
I'm playing devil's advocate and deliberately trying to find weird corner cases, so this might not apply, but...

What if you have two `device_global` variables, and one of them comes in via something like a `.so` that is dynamically linked (i.e., with an explicit call to `dlopen`) at runtime?  I presume that the initialization of the `device_global` in the `.so` wouldn't happen until it was loaded, in which case the copy constructor could end up executing after the value of the first `device_global` was modified.

Even if behavior like this is forbidden right now, it's the sort of thing I can imagine somebody asking for.

Reply:
It's not forbidden, but we could make it clear in the description of the ctor that it will be initialized to the same value as the initial value of the storage in `other`. That way, even if the value of `other` was modified, the current behavior (and in turn the behavior that we can support with minimal complexity) stays consistent.

Reply:
Instead of saying that it will be initialized to the same value as the initial value of the storage in `other`, can we say that if `other` has been modified before the constructor is executed then the behavior is undefined?

That would give us the freedom to implement whatever is easiest on different backends, rather than trying to enforce this initialization behavior on devices where it might be difficult.  It would also make it easier to change the implementation later if we decide we *do* care about the `dlopen` case, because there wouldn't be anybody relying on the behavior of our current implementation.

Reply:
Sure! I've made the changes in https://github.com/intel/llvm/pull/15075/commits/eca9cad058faacf5f8cfd4ab23518e3b82070298.